helix-core/src/indent.rs

```diff
@@ -186,106 +190,334 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
+            // (which we don't know yet)
+            first_in_line.push(None);
+        }
+        if let Some(parent) = node.parent() {
+            node = parent;
         } else {
             break;
         }
     }
 
-    Some(node)
+    let mut result = Vec::with_capacity(first_in_line.len());
+    let mut parent_is_first = true; // The root node is by definition the first node in its line
+    for first in first_in_line.into_iter().rev() {
+        if let Some(first) = first {
+            result.push(first);
+            parent_is_first = first;
+        } else {
+            result.push(parent_is_first);
+        }
+    }
+    result
 }
 
-/// Calculate the indentation at a given treesitter node.
-/// If newline is false, then any "indent" nodes on the line are ignored ("outdent" still applies).
-/// This is because the indentation is only increased starting at the second line of the node.
-fn calculate_indentation(
-    query: &IndentQuery,
-    node: Option<Node>,
-    line: usize,
-    newline: bool,
-) -> usize {
-    let mut increment: isize = 0;
+/// The total indent for some line of code.
+/// This is usually constructed by successively adding instances of [AddedIndent]
+struct IndentResult {
+    /// The total indent (the number of indent levels).
+    /// The string that this results in depends on the indent style (spaces or tabs, etc.)
+    indent: i32,
+}
+impl IndentResult {
+    fn new() -> Self {
+        IndentResult { indent: 0 }
+    }
+    /// Add the given [AddedIndent] to the [IndentResult].
+    /// The [AddedIndent] should be the combination of all the added indents for one line.
+    fn add(&mut self, added: &AddedIndent) {
+        if added.indent && !added.outdent {
+            self.indent += 1;
+        } else if added.outdent && !added.indent {
+            self.indent -= 1;
+        }
+    }
+    fn as_string(&self, indent_style: &IndentStyle) -> String {
+        indent_style.as_str().repeat(0.max(self.indent) as usize)
+    }
+}
 
-    let mut node = match node {
-        Some(node) => node,
-        None => return 0,
-    };
+/// The indent that is added for a single line.
+/// This is different from the total indent ([IndentResult]) because multiple indents/outdents on the same line don't stack.
+/// It should be constructed by successively adding the relevant indent captures.
+struct AddedIndent {
+    indent: bool,
+    outdent: bool,
+}
+impl AddedIndent {
+    /// Returns the [AddedIndent] for a line without any indent definitions
+    fn new() -> Self {
+        AddedIndent {
+            indent: false,
+            outdent: false,
+        }
+    }
+    /// Adds an indent capture to this indent.
+    fn add_capture(&mut self, capture: &IndentCaptureType) {
+        match capture {
+            IndentCaptureType::Indent => {
+                self.indent = true;
+            }
+            IndentCaptureType::Outdent => {
+                self.outdent = true;
+            }
+        }
+    }
+}
```

Comment:
With the proposal above you could then drop AddedIndent altogether and use a single `Indent` type

`Indent { indent: 1, outdent: 0 }` etc